### PR TITLE
Add reworked tests for VAT Payment Deadlines

### PR DESCRIPTION
**Pull request by gclssvglx**
[Trello](https://trello.com/c/ZiLl9eOw/627-smart-answer-flow-tests-vat-payment-deadlines)

This converts the VAT Payment Deadlines tests to use the node based testing
approach introduced in [ADR-3][1].

[1]: https://github.com/alphagov/smart-answers/blob/main/docs/arch/003-testing-flows.md

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

### CODE REVIEW

File: test/flows/vat_payment_deadlines_flow_test.rb [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    should "render the question" do
+      assert_rendered_question
+    end
+
+    context "next_node" do
+      should "have a next node of how_do_you_want_to_pay? for a date of '2013-04-30' response" do
+        assert_next_node :how_do_you_want_to_pay?, for_response: "2013-04-30"
+      end
+
+      context "validation" do
```
Comment (kevindew) [inline test/flows/vat_payment_deadlines_flow_test.rb:29]: this validation block is conventionally be outside of the `next_node` context and before it. As per: https://github.com/alphagov/smart-answers/blob/8d4d04d61c3ede76358fb10cf2470e94d4fa68a1/test/flows/state_pension_age_flow_test.rb#L35-L47

File: test/flows/vat_payment_deadlines_flow_test.rb [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+  context "question: when_does_your_vat_accounting_period_end?" do
+    setup { testing_node :when_does_your_vat_accounting_period_end? }
+
+    should "render the question" do
+      assert_rendered_question
+    end
+
+    context "next_node" do
+      should "have a next node of how_do_you_want_to_pay? for a date of '2013-04-30' response" do
```
Comment (kevindew) [inline test/flows/vat_payment_deadlines_flow_test.rb:25]:
```suggestion
      should "have a next node of how_do_you_want_to_pay? for a date of valid response" do
```

Specifying the response in the description creates the misleading impression that that response influences the next node.

File: test/flows/vat_payment_deadlines_flow_test.rb [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        assert_next_node :how_do_you_want_to_pay?, for_response: "2013-04-30"
+      end
+
+      context "validation" do
+        should "be invalid if non end-of-month date" do
+          assert_invalid_response "2013-05-15"
+        end
+
+        should "be valid if a leap year is given" do
+          assert_next_node :how_do_you_want_to_pay?, for_response: "2012-02-29"
```
Comment (kevindew) [inline test/flows/vat_payment_deadlines_flow_test.rb:35]:
```suggestion
          assert_valid_response "2012-02-29"
```

File: test/flows/vat_payment_deadlines_flow_test.rb [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        end
+
+        should "be invalid if a non leap year is given" do
+          assert_invalid_response "2011-02-29"
+        end
+      end
+    end
+  end
+
+  context "question: :how_do_you_want_to_pay?" do
```
Comment (kevindew) [inline test/flows/vat_payment_deadlines_flow_test.rb:45]:
```suggestion
  context "question: how_do_you_want_to_pay?" do
```

**Review by kevindew — COMMENTED** [review]
Looks good, just a few nitpicky things to resolve.

File: test/flows/vat_payment_deadlines_flow_test.rb (hunk 3 above)
Comment (gclssvglx) [reply test/flows/vat_payment_deadlines_flow_test.rb:35]: Following this - and the previous - suggestion this block now looks a bit odd...

```ruby
    context "next_node" do
      should "have a next node of :how_do_you_want_to_pay? for a valid response" do
        assert_valid_response: "2013-04-30"
      end
    end
```

We have a `context "next_node"` that is no longer asserting on the next node. The block looks like a validation. Are you happy this change is what you want?

**Comment by gclssvglx** [comment]
> Looks good, just a few nitpicky things to resolve.

Thanks @kevindew 👍  All fixed up.

File: test/flows/vat_payment_deadlines_flow_test.rb [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    WebMock.stub_request(:get, WorkingDays::BANK_HOLIDAYS_URL)
+      .to_return(body: File.open(fixture_file("bank_holidays.json")))
+    testing_flow VatPaymentDeadlinesFlow
+  end
+
+  should "render a start page" do
+    assert_rendered_start_page
+  end
+
+  context "question: :when_does_your_vat_accounting_period_end?" do
```
Comment (kevindew) [inline test/flows/vat_payment_deadlines_flow_test.rb:17]:
```suggestion
  context "question: when_does_your_vat_accounting_period_end?" do
```

File: test/flows/vat_payment_deadlines_flow_test.rb [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        assert_valid_response "2012-02-29"
+      end
+
+      should "be invalid if a non leap year is given" do
+        assert_invalid_response "2011-02-29"
+      end
+    end
+
+    context "next_node" do
+      should "have a next node of :how_do_you_want_to_pay? for a valid response" do
```
Comment (kevindew) [inline test/flows/vat_payment_deadlines_flow_test.rb:39]:
```suggestion
      should "have a next node of how_do_you_want_to_pay? for a valid response" do
```

File: test/flows/vat_payment_deadlines_flow_test.rb [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+      testing_node :how_do_you_want_to_pay?
+      add_responses when_does_your_vat_accounting_period_end?: "2013-04-30"
+    end
+
+    should "render the question" do
+      assert_rendered_question
+    end
+
+    context "next_node" do
+      should "have a next node of :result_direct_debit" do
```
Comment (kevindew) [inline test/flows/vat_payment_deadlines_flow_test.rb:56]:
```suggestion
      should "have a next node of result_direct_debit" do
```

**Review by kevindew — COMMENTED** [review]
Thanks, I noticed you didn't change https://github.com/alphagov/smart-answers/pull/5602#discussion_r714996416 and then I noticed there quite a few more instances of it (I didn't list them all).

Note as per https://github.com/alphagov/smart-answers/blob/main/docs/arch/003-testing-flows.md the convention is not to include a `:` prefix on node names.

**Comment by kevindew** [comment]
I noticed an error in the ADR too if you wouldn't mind reviewing 🙏  https://github.com/alphagov/smart-answers/pull/5618

**Comment by gclssvglx** [comment]
> https://github.com/alphagov/smart-answers/blob/main/docs/arch/003-testing-flows.md

Sorry, my confusion.  Hopefully, all fixed now.  Thanks

**Review by kevindew — APPROVED** [review]

**Pull request merged**